src/interpreter.rs

```diff
@@ -69,3 +72,405 @@ bitflags::bitflags! {
         const CHECKLOCKTIMEVERIFY = 1 << 9;
     }
 }
+
+fn check_minimal_push(data: &[u8], raw_opcode: u8) -> bool {
+    todo!()
+}
+
+pub fn evaluate(
+    stack: &mut Vec<Vec<u8>>,
+    script: &Script,
+    options: &VerificationFlags,
+) -> Result<(), ScriptError> {
+    // There's a limit on how large scripts can be.
+    if script.0.len() > MAX_SCRIPT_SIZE {
+        return Err(ScriptError::ScriptSize);
+    }
+
+    let mut script = (*script).clone();
+    let mut push_data = vec![];
+
+    // We keep track of how many operations have executed so far to prevent
+    // expensive-to-verify scripts
+    let mut op_count = 0;
+
+    // This keeps track of the conditional flags at each nesting level
+    // during execution. If we're in a branch of execution where *any*
+    // of these conditionals are false, we ignore opcodes unless those
+    // opcodes direct control flow (OP_IF, OP_ELSE, etc.).
+    let mut exec: Vec<bool> = vec![];
+
+    let mut alt_stack: Vec<Vec<u8>> = vec![];
+
+    // Main execution loop
+    while !script.0.is_empty() {
+        // Are we in an executing branch of the script?
+        let executing = exec.iter().all(|value| *value);
+
+        // Consume an opcode
+        let operation = parse_opcode(&mut script.0, Some(&mut push_data))?;
+
+        match operation {
+            Operation::PushBytes(raw_opcode) => {
+                // There's a limit to the size of the values we'll put on
+                // the stack.
+                if push_data.len() > MAX_SCRIPT_ELEMENT_SIZE {
+                    return Err(ScriptError::PushSize);
+                }
+
+                if executing {
+                    // Data is being pushed to the stack here; we may need to check
+                    // that the minimal script size was used to do so if our caller
+                    // requires it.
+                    if options.contains(VerificationFlags::MinimalData)
```

Comment:
Do you mean because it’s not a `const` with bitwise ops?

I can push this to a later PR and swing `VerificationFlags` back to a bunch of `const`s. I think I just wrote this bit before I decided to stick as close to the C++ as possible.

Reply:
Ah, I see. This is because of the commit it’s in, which is intended to just rearrange @ebfull’s implementation a bit.